### PR TITLE
PLAT-329: isolation negotiation

### PR DESCRIPTION
**Pull request by sexyhamster**
negotiate isolation between calls

### CODE REVIEW

File: ledger-core/v2/virtual/execute/execute.go [hunk 1]
```diff
@@ -221,10 +225,54 @@ func (s *SMExecute) stepWaitObjectReady(ctx smachine.ExecutionContext) smachine.
...
+			if err != nil {
+				panic(throw.NotImplemented())
+			}
+			s.methodIsolation = &isolation
+		}).DelayedStart().Sleep().ThenRepeat()
+	}
+
+	negotiatedIsolation, err := negotiateIsolation(*s.methodIsolation, s.execution.Isolation)
+	if err != nil {
+		ctx.Error(err)
```
Comment (bigbes) [inline ledger-core/v2/virtual/execute/execute.go:240]: return ctx.Error(err)?

Comment (sexyhamster) [reply ledger-core/v2/virtual/execute/execute.go]: thanks, fixed

File: ledger-core/v2/insolar/contract/virtual.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+)
+
+func (s PendingState) Equal(other PendingState) bool {
+	return s == other
+}
+
+type RequestResultType uint8
+
+const (
+	RequestSideEffectNone RequestResultType = iota
```
Comment (cyraxred) [inline ledger-core/v2/insolar/contract/virtual.go:69]: RequestResultType -> SideEffectType
RequestSideEffectXXX -> SideEffectXXX

Comment (sexyhamster) [reply ledger-core/v2/insolar/contract/virtual.go]: fixed

File: ledger-core/v2/application/preprocessor/main.go [hunk 3]
```diff
@@ -470,8 +472,9 @@ func (pf *ParsedFile) functionInfoForWrapper(list []*ast.FuncDecl) []map[string]
 			"ResultDefinitions":   numberedDefinition(pf, fun.Type.Results, "ret"),
 			"ErrorInterfaceInRes": errorInterfaceInRes,
 			"LastErrorInRes":      lastErrorInRes,
-			"Immutable":           isImmutable(fun),  // only for methods, not constructors
-			"SagaInfo":            sagaInfo(pf, fun), // only for methods, not constructors
+			"Interference":        geInterference(fun), // only for methods, not constructors
```
Comment (bigbes) [inline ledger-core/v2/application/preprocessor/main.go:475]: gege

File: ledger-core/v2/insolar/contract/virtual.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+)
+
+func (i PendingState) Equal(other PendingState) bool {
+	return i == other
+}
+
+type SideEffectType uint8
+
+const (
+	SideEffectTypeNone SideEffectType = iota
```
Comment (bigbes) [inline ledger-core/v2/insolar/contract/virtual.go:69]: maybe we should move that declaration to separate package and remove prefix SideEffect?
If not, then rename `SideEffectTypeNone ` -> `SideEffectNone `

Comment (bigbes) [reply ledger-core/v2/insolar/contract/virtual.go:69]: As a side effect of moving to another package we can `go gen` Stringer

File: ledger-core/v2/runner/adapter.go [hunk 5]
```diff
@@ -127,15 +128,17 @@ func (a *ServiceAdapter) PrepareExecutionAbort(ctx smachine.ExecutionContext, st
 	})
 }
 
-func (a *ServiceAdapter) PrepareExecutionClassify(ctx smachine.ExecutionContext, execution execution.Context, fn func(interface{})) smachine.AsyncCallRequester {
+func (a *ServiceAdapter) PrepareExecutionClassify(ctx smachine.ExecutionContext, execution execution.Context, fn func(contract.MethodIsolation, error)) smachine.AsyncCallRequester {
 	if fn == nil {
 		panic(throw.IllegalValue())
 	}
+	proto, _ := execution.ObjectDescriptor.Prototype()
```
Comment (bigbes) [inline ledger-core/v2/runner/adapter.go:135]: debug log?

File: ledger-core/v2/runner/executionevent/event.go [hunk 6]
```diff
@@ -181,13 +186,12 @@ func (e CallMethod) ConstructVCallRequest(execution execution.Context) *payload.
 	execution.Sequence++
```
Comment (bigbes) [inline ledger-core/v2/runner/executionevent/event.go:186]: Just saw that execution.Context is not pointer, but object :(
It'll lead to worthlessness increment.

File: ledger-core/v2/runner/executor/builtin/builtin.go [hunk 7]
```diff
@@ -94,21 +95,26 @@ func (r *Runner) CallMethod(
...
+	return methodObject.Func(data, args)
+}
+
+func (r *Runner) ClassifyMethod(_ context.Context,
+	codeRef reference.Global,
+	method string) (contract.MethodIsolation, error) {
+
+	contractName, ok := r.CodeRefRegistry[codeRef]
+	if !ok {
+		return contract.MethodIsolation{},
```
Comment (bigbes) [inline ledger-core/v2/runner/executor/builtin/builtin.go:107]: Isn't it better to split that return to 2 step?
```
err := throw.E("failed to find contract with reference", struct{ Reference reference.Global }{Reference: codeRef})
return contract.MethodIsolation{}, err
```

File: ledger-core/v2/insolar/contract/virtual.go (hunk 4 above)
Comment (sexyhamster) [reply ledger-core/v2/insolar/contract/virtual.go:69]: I'm not sure that we even need this, both SideEffect and PendingState, probably will just delete all of them, as they are not used

File: ledger-core/v2/runner/executionevent/event.go (hunk 6 above)
Comment (sexyhamster) [reply ledger-core/v2/runner/executionevent/event.go:186]: should be fixed in different PR with test that proves it >_<

File: ledger-core/v2/insolar/contract/virtual.go (hunk 4 above)
Comment (bigbes) [reply ledger-core/v2/insolar/contract/virtual.go:69]: SideEffectType was used, as i remember, when we determine what should we do with the result of execution, or am i wrong?

File: ledger-core/v2/runner/executionevent/event.go (hunk 6 above)
Comment (bigbes) [reply ledger-core/v2/runner/executionevent/event.go:186]: sure

File: ledger-core/v2/insolar/contract/virtual.go (hunk 4 above)
Comment (sexyhamster) [reply ledger-core/v2/insolar/contract/virtual.go:69]: it is different side effect type defined in requestresult package :(

File: ledger-core/v2/runner/adapter.go (hunk 5 above)
Comment (sexyhamster) [reply ledger-core/v2/runner/adapter.go]: removed

File: ledger-core/v2/runner/executor/builtin/builtin.go (hunk 7 above)
Comment (sexyhamster) [reply ledger-core/v2/runner/executor/builtin/builtin.go]: fixed

File: ledger-core/v2/application/preprocessor/main.go (hunk 3 above)
Comment (sexyhamster) [reply ledger-core/v2/application/preprocessor/main.go]: fixed

**Pull request merged**